nvaccess/nvda
# Native driver for Freedom Scientific displays

Edited by @leonardder

### Link to issue number:
Fixes #7727
fixes #4464
fixes #8849 
fixes #8729

### Summary of the issue:

The current Freedom Scientific braille display driver requires a binary component from Freedom Scientific. This prevents the driver from integrating fully with NVDA features like braille display auto detection. Also, braille input is not supported by the current driver. 

### Description of how this pull request fixes the issue:
1. This is a rewrite of the driver in pure Python. It still requires the USB driver from Freedom Scientific for working with displays over USB, but the communication with the display is now handled in Python.

    It also supports braille input and auto detection.

2. This PR adds a new feature to braille auto detection as well. It keeps searching for a USB display when a Bluetooth connection is active, it will only connect to USB devices for the same driver. This is because a Focus display switches to USB when a cable is plugged in, even if a Bluetooth connection is active. The old binary driver does auto switching to USB as well, so not supporting this would be a regression for users. This also works for other display brands.

### Testing performed:

* Tested with a Focus 40 blue 4th generation over USB and Bluetooth by me
* Tested with a Focus 40 fifth generation by @MarcoZehe over USB and Bluetooth
* Other users have done tests as well during development of this driver, see #7727 

* @LeonarddeR tested the new bdDetect behavior as follows:

	1. Connect a Handy Tech Basic Braille via Bluetooth. The display is detected
	2. Connect a Handy Tech Easy Braille via USB. The Basic Braille is automatically disconnected in favour of the USB device.
	3. Disconnect the Easy Braille. NVDA switches back to the Basic Braille over bluetooth automatically.
	4. Connect the Handy Tech Easy Braille via USB again. The Basic Braille is automatically disconnected in favour of the USB device.
	5. Again, disconnect the Easy Braille. NVDA switches back to the Basic Braille over bluetooth automatically.
	6. Connect an Eurobraille Esys over USB. NVDA keeps the connection to the Basic Braille over Bluetooth, since the Esys doesn't belong to the Handy Tech driver

### Known issues with pull request:

* None

### Change log entry:

* New features
  * Freedom Scientific braille displays are now supported by braille display auto detection (#7727)

* Changes
  * When an auto detected braille display is connected via Bluetooth, NVDA will keep searching for USB displays supported by the same driver and switch to a USB connection if it becomes available

* Bugfixes
  * The bumper keys now work correctly on Freedom Scientific braille displays (#8849)
